delpay: introduced a new rpc method to delete a payment by hash

vincenzopalazzo opened the pull request:
Hi all.

This Draft implement a new RPC command to delete the payment with the status `complete` `failed` marked like good first issue https://github.com/ElementsProject/lightning/issues/1995

However, this PR depends from the PR where I proposed to update the result of listpays command with `payment_hash` inside the result https://github.com/ElementsProject/lightning/pull/3888. In addition, with the @cdecker suggestion to add the `payment_hash` in all cases of types of payments, the command `delpay` can use only the `payment_hash` to identify the payment to remove and not `bolt11`.

This PR is a draft yet because I have some point to clarify with the team, for instance, the type of error that the command should be returned in cases of error. In other words, I see that the `LIGHTNINGD` error is marked with a `FIXME` because each command should use the correct error code. I'm not able to find the payment error codes inside the code. maybe we could define the payment error code?

- [x] Docs update with the change proposed in this Draft
- [X] Fix error code inside the command implementation

Changelog-Added: JSON-RPC: delpay a new method to delete the payment completed or failed.

ZmnSCPxj commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,58 @@
+SYNOPSIS
+--------
+
+**delpay** *payment_hash* \[status\] (is not specify the status is complete)
```
You have a `partid` as well that is not documented.

Defaults should be described in "DESCRIPTION" section, not in the synopsis.

vincenzopalazzo replied:
Updaed

ZmnSCPxj commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,58 @@
+
+The **delpay** RPC command removes a payment as given in **listsendpays** or **listpays** with a complete or failed
+status. However, the command doesn't permit to remove the pending payment.
+
```
DESCRIPTION section needs to describe every parameter, and describe defaults for that parameter as well if it is optional and not specified.

vincenzopalazzo replied:
Updated but the doc needs a second review in the next hours (Caused from my Dyslexia problem :smile: )

ZmnSCPxj commented:
In addition, it seems to me better to just delete *all* `partid`s, not delete them one-by-one.  Loss of one `partid` of a multi-part payment is already a loss of information anyway, so just delete them all in one database transaction.

vincenzopalazzo commented:
Hi @ZmnSCPxj, thanks to review this PR.

I will make the changes inside the docs and inside the JSON RPC errors, sorry about that.

In addition, about your suggestion.

>In addition, it seems to me better to just delete all partid, not delete them one-by-one. Loss of one partid of a multi-part payment is already a loss of information anyway, so just delete them all in one database transaction.

For my code comprehension, to make this work, so, to delete the complete payment and not each payment by `partid`, I need a new function inside the wallet.c to run the query "DELETE FROM payments WHERE payment_hash=?"

I'm losing something?

Maybe I can use the `function wallet_payment_list` and run for each element inside the list `wallet_payment_delete` but I think that the first solution is cleaner but I don't know if it is safe to run on the lightning node the delete query by `hash`?

ZmnSCPxj commented:
`DELETE FROM TABLE payments WHERE payment_hash=?` should be fine.  You might want to do some check to check if this results in any changes (I think we have a function to check the number of changes a query makes) and if it is 0 that is probably fine.

But you want to check the state of the *entire* payment first. The logic for a purported "payment state":

* If any part has succeeded, then the entire payment succeeded.
* If any part is pending, then the entire payment is pending.
* Otherwise, all parts are failed, so the entire payment is failed.

It is safe to delete a payment if the entire payment is failed or succeeded, and we should probably disallow deleting pending payments (which might still be operated by `pay` over in the `pay` plugin).

Note that if you do not pass control to a function with a callback, then the entire processing is done inside a database transaction, so you are assured that between the check and the deletion, the database state is consistent.  But if you pass control to a function with a callback, the database transaction will end at that point, and by the time the callback is called, you are in a new transaction.  So do not call into any callback-using functions between checking the state of the payment and deleting the whole payment.

vincenzopalazzo commented:
Thanks for the suggestion @ZmnSCPxj.

>You might want to do some check to check if this results in any changes (I think we have a function to check the number of changes a query makes) and if it is 0 that is probably a "payment not found" error.

I make the sanity test of input before to call the database function, but I can improve the check maybe I can make the following test:

1. payment not found: I used the `wallet_payment_by_hash` to find the payment, I think that is possible use this function to get the payment with hash and *partid=0*, if exist this part, the payment exists, right?

2. payment status not valid: from your previous message I understand the payment status is decided from the status of all part, so, I should check the status of all payment find with `wallet_payment_list` and check the status for each payment and make the correct decision.

of course, there are all checks on the input but, they are not relevant into this description.

>It is safe to delete a payment if the entire payment is failed or succeeded, and we should probably disallow deleting pending payments (which might still be operated by pay over in the pay plugin).

Yep, I agree, with my actual logic inside the `delpay` command you can remove only the payment with the status `complete` or `failed`, but of course, with your suggestion the logic change a little.

ZmnSCPxj commented:
`partid` 0 is for non-split payments.  If we split payments at all, there might not be a `partid` 0.

vincenzopalazzo commented:
Rebased https://github.com/ElementsProject/lightning/commit/01a82d38f769a410e4f2d127c9f09e8f8b296554 and updated https://github.com/vincenzopalazzo/lightning/commit/3a5f0be11e8f9efe6bc7796a0fd3289e83aebc52

ZmnSCPxj commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,83 @@
+status. However, the command doesn't permit to remove a pending payment.
+
+- *payment_hash*: Rapresents the unique identifier of a payment. To find it, the caller can run **listpays** or **listsendpays**;
+- *status*: Represents a payment status. Otherwise, if it is not specified it is equal to *complete*.
```
You should list here the possible values.

```
*status* is the expected status of the payment.
It can be *complete* or *pending*.
Only delete if the payment status matches.
If not specified, defaults to *complete*.
```

vincenzopalazzo replied:
Thanks for this suggestion. modify made into https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

ZmnSCPxj commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,83 @@
+The following error codes may occur:
+
+- -32602: Some parameter missed or some parameter is malformed;
+- 211: Payment with payment_hash have a wrong status. To check the correct status run the command **paystatus**;
```
```
*payment\_hash*
```

vincenzopalazzo replied:
Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

ZmnSCPxj commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,83 @@
+SYNOPSIS
+--------
+
+**delpay** *payment_hash* \[status\]
```
Escape the underscore (as well all instances)

```
*payment\_hash*
```

vincenzopalazzo replied:
Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

ZmnSCPxj commented on lightningd/pay.c:
```diff
@@ -1496,6 +1521,77 @@ static const struct json_command listsendpays_command = {
+		NULL))
+		return command_param_failed();
+
+	if (!payment_hash)
```
Check is unnecessary, `p_req` will check already.

vincenzopalazzo replied:
Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

ZmnSCPxj commented on lightningd/pay.c:
```diff
@@ -1496,6 +1521,77 @@ static const struct json_command listsendpays_command = {
+
+	if (!param(cmd, buffer, params,
+		p_req("payment_hash", param_sha256, &payment_hash),
+		p_opt("status", param_string, &status_str),
```
Can use instead:

```c
p_opt_def("status", param_string, &status_str, "complete")
```

vincenzopalazzo replied:
@ZmnSCPxj, mh! maybe I'm losing something here, param_string accept only a `const char *str` and the function `o_opt_def` can not assign a new value at `status_str` variable.

With this comment, you mean that I need to declare some function that accepts a not const `char*`?

I think that I'm losing something, sorry about that

ZmnSCPxj replied:
String constants *are* `const char *`.  You can do something like `const char *foo = "foo";`.  You can even do something like `const char *foo = "foo"; foo = "bar"`.

The *correct* way to write `const` is *after* the type.  So actually `const char *` should be written as `char const *`.  Then you read types from right to left.  So `char const *` is read as "pointer (`*`) to constant (`const`) character (`char`)".  Note that the pointer itself is not `const`, so you can modify the pointer, which is why `char const* foo; foo = "bar"; foo = "quux"` is valid.  What you cannot do is `char const* foo = "foo"; *foo = '0';`, because what is pointed *to* is constant.

vincenzopalazzo replied:
Thanks for this wonderful informations. I made a couple of change inside https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934 but I want to have more detail about these changes. I like your suggestion to use p_opt_def.

ZmnSCPxj commented on lightningd/pay.c:
```diff
@@ -1496,6 +1521,77 @@ static const struct json_command listsendpays_command = {
+						"missing required parameter: payment_hash");
+
+
+	status = status_str == NULL ? PAYMENT_COMPLETE : string_to_payment_status(status_str);
```
See above for `p_opt_def`.

ZmnSCPxj commented on lightningd/pay.c:
```diff
@@ -1496,6 +1521,77 @@ static const struct json_command listsendpays_command = {
+
+	if (status == -1)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						"status insert %s wrong", status_str);
```
"status insert"?  Better "Unrecognized status: %s".

vincenzopalazzo replied:
Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

ZmnSCPxj commented on lightningd/pay.c:
```diff
@@ -42,6 +42,31 @@ struct sendpay_command {
 	struct command *cmd;
 };
 
+static enum wallet_payment_status string_to_payment_status(const char *status_str)
```
The usual convention for parsers that can fail is something like this:

```c
bool string_to_payment_status(const char *status_str, enum wallet_payment_status *)
```

Then later in use:

```
if (!string_to_payment_status(status_str, &status))
    return command_fail(cmd, JSONRPC2_INVALID_PARAMS, "Unrecognized status '%s'", status_str);
```

vincenzopalazzo replied:
Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

ZmnSCPxj commented on wallet/wallet.c:
```diff
@@ -2682,6 +2682,17 @@ void wallet_payment_delete(struct wallet *wallet,
+{
+	struct db_stmt *stmt;
+	stmt = db_prepare_v2(
+		wallet->db, SQL("DELETE FROM payments WHERE payment_hash = ?"));
```
Looks correct but @cdecker should check as well, I am no longer familiar with how we interact with the db.

cdecker replied:
This should work just fine. The SQL subsystem just takes a valid query and rewrites it for various dialects. That works well for this type of simple queries, but may be a bit more complex for things like upserts or last inserted IDs, which we use in some places.

ZmnSCPxj commented on lightningd/pay.c:
```diff
@@ -1496,6 +1521,77 @@ static const struct json_command listsendpays_command = {
+						"status insert %s wrong", status_str);
+
+	if (status == PAYMENT_PENDING)
+		return command_fail(cmd, PAY_STATUS_UNEXPECTED, "invalid status: %s",
```
This is an error in the argument, not in the status of the actual payment, so should be `JSONRPC2_INVALID_PARAMS`.

vincenzopalazzo replied:
Changed in https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

vincenzopalazzo commented on lightningd/pay.c:
```diff
@@ -42,6 +42,34 @@ struct sendpay_command {
 	struct command *cmd;
 };
 
+static bool string_to_payment_status(const char *status_str, enum wallet_payment_status *status)
```
@ZmnSCPxj I understood your suggestions inside this function? or I'm introducing some mistake?

ZmnSCPxj replied:
looks ok, but no time for in-depth review now, please wait....

niftynei commented on tests/test_pay.py:
```diff
@@ -3242,6 +3242,107 @@ def test_mpp_presplit_routehint_conflict(node_factory, bitcoind):
+    invl1 = l1.rpc.invoice(Millisatoshi(amount_sat * 2 * 1000), "j", "j")
+    l2.rpc.pay(invl1["bolt11"])
+
+    before_del_pay = l2.rpc.listpays()
```
this test could easily be rolled into the previous one

niftynei commented on lightningd/pay.c:
```diff
@@ -1498,6 +1526,73 @@ static const struct json_command listsendpays_command = {
+
+	if (tal_count(payments) == 0)
+		return command_fail(cmd, PAY_NO_SUCH_PAYMENT,
+			"Unknown payment with payment_hash: %s",
```
line this up with `cmd` on the previous line?

niftynei commented on lightningd/pay.c:
```diff
@@ -1498,6 +1526,73 @@ static const struct json_command listsendpays_command = {
+
+	if (!string_to_payment_status(status_str, &status))
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						"Unrecognized status: %s", status_str);
```
needs to be aligned with `cmd`

niftynei commented on lightningd/pay.c:
```diff
@@ -1498,6 +1526,73 @@ static const struct json_command listsendpays_command = {
+	for (int i = 0; i < tal_count(payments); i++) {
+		if (payments[i]->status != status) {
+			return command_fail(cmd, PAY_STATUS_UNEXPECTED,
+					"Payment with hash %s has %s status but it should be %s",
```
align with `cmd`?

niftynei commented on lightningd/pay.c:
```diff
@@ -42,6 +42,34 @@ struct sendpay_command {
+		return "complete";
+	case PAYMENT_FAILED:
+		return "failed";
+	default:
```
it's best practice not to use `default` for a switch statement. instead, name all of the cases.

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+-----------
+
+The **delpay** RPC command removes a payment as given in **listsendpays** or **listpays** with a complete or failed
+status. However, the command doesn't permit to remove a pending payment.
```
```suggestion
status. Deleting a `pending` payment is an error.
```

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+The **delpay** RPC command removes a payment as given in **listsendpays** or **listpays** with a complete or failed
+status. However, the command doesn't permit to remove a pending payment.
+
+- *payment\_hash*: Rapresents the unique identifier of a payment. To find it, you can run **listpays** or **listsendpays**;
```
```suggestion
- *payment\_hash*: The unique identifier of a payment. Find in **listpays** or **listsendpays**;
```

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+status. However, the command doesn't permit to remove a pending payment.
+
+- *payment\_hash*: Rapresents the unique identifier of a payment. To find it, you can run **listpays** or **listsendpays**;
+- *status* is the expected status of the payment. It can be *complete* or *failed*. 
```
```suggestion
- *status*: Expected status of the payment. Valid values are *complete* or *failed*. 
```

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+
+- *payment\_hash*: Rapresents the unique identifier of a payment. To find it, you can run **listpays** or **listsendpays**;
+- *status* is the expected status of the payment. It can be *complete* or *failed*. 
+Only delete if the payment status matches. If not specified, defaults to *complete*.
```
```suggestion
Only deletes if the payment status matches. If not specified, defaults to *complete*.
```

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+RETURN VALUE
+------------
+
+On success, the command will return a payment object, such as the **listsendpays**. In addition, if the payment is a MPP (Multi part payment) the command return a list of 
```
```suggestion
On success, the command will return a payment object, such as the **listsendpays**. If the payment is an MPP (multi-part payment) a list of 
```

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+------------
+
+On success, the command will return a payment object, such as the **listsendpays**. In addition, if the payment is a MPP (Multi part payment) the command return a list of 
+payments; a payment object for each partid.
```
```suggestion
payments will be returned -- one payment object for each partid.
```

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+On success, the command will return a payment object, such as the **listsendpays**. In addition, if the payment is a MPP (Multi part payment) the command return a list of 
+payments; a payment object for each partid.
+
+On failure, an error is returned and any payment is deleted. If the lightning process fails before responding, the
```
```suggestion
On failure, an error is returned. If the lightning process fails before responding, the
```

we don't delete the payment on failure?

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+
+The following error codes may occur:
+
+- -32602: Some parameter missed or some parameter is malformed;
```
```suggestion
- -32602: Parameter missing or malformed;
```

niftynei commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+The following error codes may occur:
+
+- -32602: Some parameter missed or some parameter is malformed;
+- 211: Payment with payment\_hash have a wrong status. To check the correct status run the command **paystatus**;
```
```suggestion
- 211: Payment status mismatch. Check the correct status via **paystatus**;
```

niftynei commented:
Mostly grammatical/spelling fixes

vincenzopalazzo commented:
Rebase and update with suggestions.

Thanks @niftynei about the fixes and sorry for the spelling errors. I'm still searching somethings where I'm good to do but I'm sure that I'm very bad to write sentences :smile:

cdecker commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+DESCRIPTION
+-----------
+
+The **delpay** RPC command removes a payment as given in **listsendpays** or **listpays** with a complete or failed
+status. Deleting a `pending` payment is an error.
```
```suggestion
The **delpay** RPC command deletes a payment with the given `payment_hash` if its status is either `complete` or `failed`. Deleting a `pending` payment is an error.
```

cdecker commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+
+- *payment\_hash*: The unique identifier of a payment. Find it, you can run **listpays** or **listsendpays**;
+- *status*: Expected status of the payment. Valid values are *complete* or *failed*. 
+Only deletes if the payment status matches. If not specified, defaults to *complete*.
```
Shouldn't an unspecified value for `status` mean we don't care, and either `complete` or `failed` will be removed?

ZmnSCPxj replied:
Since `delpay` is supposed to mirror `delinvoice`, which has the `*status*` as a required arg, maybe it should be a required arg here as well?

vincenzopalazzo replied:
Yep I agree with this point! thanks for clarifying!

cdecker commented on doc/lightning-delpay.7.md:
```diff
@@ -0,0 +1,84 @@
+RETURN VALUE
+------------
+
+On success, the command will return a payment object, such as the **listsendpays**. If the payment is aa MPP (multi part payment) the command return a list of 
```
```suggestion
If successful the command returns a payment object, in the same format as **listsendpays**.
If the payment is a multi-part payment (MPP) the command return a list of 
```

cdecker commented on lightningd/pay.c:
```diff
@@ -1498,6 +1528,70 @@ static const struct json_command listsendpays_command = {
+	if (!string_to_payment_status(status_str, &status))
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS, "Unrecognized status: %s", status_str);
+
+	if (status == PAYMENT_PENDING)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS, "Invalid status: %s",
+							payment_status_to_string(status));
```
Probably it's better to check if we have one of the valid statuses instead of not having the third. If we ever add a new status this is would not cover that fourth state. If you use a `switch` statement then the compiler will make sure we cover all cases exhaustively.

vincenzopalazzo replied:
Thanks for this suggestion, I made the change from if to switch!

cdecker commented on wallet/wallet.c:
```diff
@@ -2669,7 +2669,7 @@ void wallet_payment_store(struct wallet *wallet,
 		db_bind_text(stmt, 10, payment->bolt11);
 	else
-		db_bind_null(stmt, 10);
+	       	db_bind_null(stmt, 10);
```
Spurious whitespace change.

vincenzopalazzo replied:
Ops! thanks

cdecker commented on wallet/wallet.h:
```diff
@@ -1026,6 +1026,15 @@ void wallet_payment_delete(struct wallet *wallet,
+ * Removes the payment from the database by hash; if it is a MPP payment
+ * it remove all parts with a single query.
+ */
+void wallet_payment_delete_by_hash(struct wallet *wallet,
+				   const struct sha256 *payment_hash);
```
Nit: indentation of the followup line should match the `struct` keyword above (might be a GH display issue though).

cdecker commented on tests/test_pay.py:
```diff
@@ -3261,6 +3261,77 @@ def test_mpp_presplit_routehint_conflict(node_factory, bitcoind):
 
+def test_delpay_argument_invalid(node_factory, bitcoind):
+    """
+    This test includes all possible combination of input error inside the
```
```suggestion
    This test includes all possible combinations of input error inside the
```

cdecker commented on tests/test_pay.py:
```diff
@@ -3261,6 +3261,77 @@ def test_mpp_presplit_routehint_conflict(node_factory, bitcoind):
+    with pytest.raises(RpcError):
+        l2.rpc.delpay()
+
+    # invoice unpaid
+    inv = l1.rpc.invoice(10 ** 5, 'inv', 'inv')
+    payment_hash = inv["payment_hash"]
+    with pytest.raises(RpcError):
+        l2.rpc.delpay(payment_hash)
```
Well, really this matches the "no payment yet" case, since `l2` doesn't know anything about that payment. So you could just use `payment_hash = "AA"*32` here.

cdecker commented on tests/test_pay.py:
```diff
@@ -3261,6 +3261,77 @@ def test_mpp_presplit_routehint_conflict(node_factory, bitcoind):
+    with pytest.raises(RpcError):
+        l2.rpc.delpay(payment_hash)
+
+    # payment unpaid with wrong status (pending status is a illegal input)
+    with pytest.raises(RpcError):
+        l2.rpc.delpay(payment_hash, 'pending')
```
Payment was no attempted at all, so it'd fail for that reason already.

cdecker replied:
Basically this is testing our sanity check order, and not the underlying `pending` vs. `non-pending` cases.

vincenzopalazzo replied:
This comment brings me to reflect on the sens of this test!

I moved the pay command to the top of this sanity check, so, with this new situation the unit test has more sense to me because it does the sanity check and the command behavior checks

cdecker commented on tests/test_pay.py:
```diff
@@ -3261,6 +3261,77 @@ def test_mpp_presplit_routehint_conflict(node_factory, bitcoind):
+    with pytest.raises(RpcError):
+        l2.rpc.delpay(payment_hash, 'invalid_status')
+
+    l2.rpc.connect(l1.info['id'], 'localhost', l1.port)
+    l2.fund_channel(l1, 10 ** 6)
```
You could have used `l1, l2 = node_factory.line_graph(2)` above since so far the channel has not been used. That'd save a couple of test-lines

vincenzopalazzo replied:
In these cases, I could be used the following line of code to have the channels ready to work `node_factory.line_graph(2, fundamount=10**5, wait_for_announce=True)` , right?

cdecker commented on tests/test_pay.py:
```diff
@@ -3261,6 +3261,77 @@ def test_mpp_presplit_routehint_conflict(node_factory, bitcoind):
+    l2.rpc.connect(l1.info['id'], 'localhost', l1.port)
+    l2.fund_channel(l1, 10 ** 6)
+
+    bitcoind.generate_block(6)
+    sync_blockheight(bitcoind, [l1, l2])
+
+    wait_for(lambda: len(l2.rpc.listchannels()['channels']) == 2)
```
These would have also been subsumed by `line_graph`

cdecker commented on tests/test_pay.py:
```diff
@@ -3261,6 +3261,77 @@ def test_mpp_presplit_routehint_conflict(node_factory, bitcoind):
+    with pytest.raises(RpcError):
+        l2.rpc.delpay(payment_hash, 'failed')
+
+    with pytest.raises(RpcError):
+        l2.rpc.delpay(payment_hash, 'pending')
```
This is redundant with the test on line 3283

cdecker commented on tests/test_pay.py:
```diff
@@ -3261,6 +3261,77 @@ def test_mpp_presplit_routehint_conflict(node_factory, bitcoind):
+
+    l1, l2, l3 = node_factory.line_graph(
+        3, fundamount=10**8, wait_for_announce=True,
+        opts={'wumbo': None}
```
Wumbo is not needed here: `MPP_TARGET_SIZE` is denominated in msats, whereas the `fundamount` is denominated in sats. So this creates a channel that is 4 orders of magnitude larger than the amount being transferred:

 - Channel capacity: 10**8 * 1000 = 10**11
 - Payment amount: 5 * 10**7

Doesn't hurt though :wink:

vincenzopalazzo replied:
>Doesn't hurt though

Yes, it is, but it is also my logic error, I discovered only now that `fundamount` is denominated in sats. I think that `10**5` is enough

vincenzopalazzo commented:
Rebase and update with review https://github.com/ElementsProject/lightning/commit/3314a059779c4769771a4d086b772b9c40827c14

Waiting to Travis test result

cdecker commented:
Rebased on top of `master`

cdecker reviewed (approved):
ACK https://github.com/ElementsProject/lightning/pull/3899/commits/56a5dc93d43bbadaa64b5be9dd0c07189f5dab27

rustyrussell commented:
Ack https://github.com/ElementsProject/lightning/pull/3899/commits/db5323522e23bb5f918ec0d96790316212758621
